Decouple client release from Delphi release

rzats opened the pull request:
Closes #1296.

### Summary:

_Semi-detaches_ the versioning of the Python client and delphi-epidata.

This means that a new version of the Python client is only released when changes are made to the client file - when this is the case, however, it is synced to the current delphi-epidata version.

For further details, see https://github.com/cmu-delphi/delphi-epidata/issues/1296#issuecomment-2143152714.

### Prerequisites:

- [x] Unless it is a documentation hotfix it should be merged against the `dev` branch
- [x] Branch is up-to-date with the branch to be merged with, i.e. `dev`
- [x] Build is successful
- [x] Code is cleaned up and formatted

rzats commented:
For testing purposes, I've forked the `delphi-epidata` repository, then added the commit https://github.com/cmu-delphi/delphi-epidata/commit/917a89b533acb0c4bfd1f368e23fae1c8cbb0bb1 to it. This is a version of this PR which enables `create-release` and `release-helper` within a fork, _but_ also stubs out various destructive steps in the GitHub Action - like actually creating a PyPi release, for instance.

I then ran several mock releases within the same repository:

**4.1.24**

https://github.com/rzats/delphi-epidata/pull/1
- Simply syncs `main` with `dev`.
- There is currently a client change in `dev` which is not in `main`, so all versions are bumped from 4.1.23 to 4.1.24.
- In the `release-helper` run, the "release Python client" step would be launched (if it wasn't stubbed out). https://github.com/rzats/delphi-epidata/actions/runs/9355684178

**4.1.25**

https://github.com/rzats/delphi-epidata/pull/3
- Adds a simple change to `README.md`.
- There are no changes to the client, so Python client versions remain unchanged, while others are bumped from 4.1.24 to 4.1.25.
- In the `release-helper` run, the "release Python client" step would not be launched. https://github.com/rzats/delphi-epidata/actions/runs/9355704159

**4.1.26**

https://github.com/rzats/delphi-epidata/pull/3
- Adds a simple change to the client changelog.
- Python client versions are correctly bumped from 4.1.24 to 4.1.26, skipping over a missing version. Others are bumped from 4.1.25 to 4.1.26.
- In the `release-helper` run, the "release Python client" step would be launched. https://github.com/rzats/delphi-epidata/actions/runs/9355727562

melange396 commented on .github/workflows/create-release.yml:
```diff
@@ -31,6 +31,22 @@ jobs:
+        run: echo "sha=$(git rev-parse origin/main)" >> $GITHUB_OUTPUT
+      - name: Get Python client file changes
+        id: changed-py
+        uses: tj-actions/changed-files@v44
```
should we use `dorny/paths-filter` here (and also in `release-helper.yml`) to match the usage in the `covidcast-indicators` repo?  fewer dependencies and reinvented wheels make for easier support and maintenance.
https://github.com/cmu-delphi/covidcast-indicators/blob/a6ea00355d79a47cc09db8090f6e5121c718de0f/.github/workflows/create-release.yml#L32-L40
https://github.com/cmu-delphi/covidcast-indicators/blob/a6ea00355d79a47cc09db8090f6e5121c718de0f/.github/workflows/publish-release.yml#L31-L41

rzats replied:
I can give it a go, it should support the same features as `changed-files`.

melange396 commented on .github/workflows/create-release.yml:
```diff
@@ -31,6 +31,22 @@ jobs:
+        if: steps.changed-py.outputs.any_changed == 'true'
+        run: |
+          cd src/client/packaging/pypi
+          bump2version --allow-dirty --new-version ${{ steps.version.outputs.next_tag }} ignore_part
```
please make a comment that notes this is using `${{ steps.version.outputs.next_tag }}`, the new literal version number output from the `bump2version` step above, not the workflow input `${{ github.event.inputs.versionName }}` which can be relative, thus ensuring the version numbers stay in sync.

it looks like `ignore_part` is just a placeholder for the otherwise required "part" field that is being overridden by `--new-version`...  can we simplify this to something closer in format to our other uses of `bump2version`?
```suggestion
          bump2version --allow-dirty ${{ steps.version.outputs.next_tag }}
```

also, is `--allow-dirty` necessary here?

rzats replied:
> please make a comment that notes this is using `${{ steps.version.outputs.next_tag }}`, the new literal version number output from the `bump2version` step above, not the workflow input` ${{ github.event.inputs.versionName }}` which can be relative, thus ensuring the version numbers stay in sync.

Will do!

> it looks like `ignore_part` is just a placeholder for the otherwise required "part" field that is being overridden by `--new-version`... can we simplify this to something closer in format to our other uses of `bump2version`?

Unfortunately no, this is a quirk of the tool: https://github.com/c4urself/bump2version/issues/22

> also, is `--allow-dirty` necessary here?

It is, because we use `bump2version` twice: first to bump the versions for other than the Python client (which makes the working directory "dirty"), then the Python client as well.

melange396 replied:
> > it looks like `ignore_part` is just a placeholder for the otherwise required "part" field that is being overridden by `--new-version`... can we simplify this to something closer in format to our other uses of `bump2version`?
> 
> Unfortunately no, this is a quirk of the tool: https://github.com/c4urself/bump2version/issues/22
> 

wow, that is disappointing on multiple levels: bump2version is arguably broken, and this line has (and the 9 other places in the org that its been copied to have too) been lying to us about acceptable inputs: https://github.com/cmu-delphi/delphi-epidata/blob/69835d1d7795eaf9a710d9f4903fef22a07e8fdf/.github/workflows/create-release.yml#L5-L7

dshemetov replied:
FWIW, bump2version says it's no longer maintained and that [this](https://github.com/callowayproject/bump-my-version) is the maintained fork.

dshemetov replied:
[Tried it](https://github.com/cmu-delphi/delphi-epidata/pull/1473) and eh, it still has issues.

dshemetov commented:
Pushed a merge from dev and resolved conflicts with #1470.

rzats commented:
@melange396 @dshemetov this is ready once again! Here's an updated rundown of my CI tests as well.

For testing purposes, I've forked the `delphi-epidata` repository, then added the commit https://github.com/cmu-delphi/delphi-epidata/commit/bf3dec4328e11474a1fa361a9b8608d3aff08a21 to it. This is a version of this PR which enables `create-release` and `release-helper` within a fork, _but_ also stubs out various destructive steps in the GitHub Action - like actually creating a PyPi release, for instance.

I then ran several mock releases within the same repository:

**4.1.24**

https://github.com/rzats/delphi-epidata/pull/1
- Simply syncs `main` with `dev`.
- There is currently a client change in `dev` which is not in `main`, so all versions are bumped from 4.1.23 to 4.1.24.
- In the `release-helper` run, the "release Python client" step would be launched (if it wasn't stubbed out). https://github.com/rzats/delphi-epidata/actions/runs/9616207616

**4.1.25**

https://github.com/rzats/delphi-epidata/pull/3
- Adds a simple change to `README.md`.
- There are no changes to the client, so Python client versions remain unchanged, while others are bumped from 4.1.24 to 4.1.25.
- In the `release-helper` run, the "release Python client" step would not be launched. https://github.com/rzats/delphi-epidata/actions/runs/9616236975

**4.1.26**

https://github.com/rzats/delphi-epidata/pull/5
- Adds a simple change to `delphi_epidata.py` within the client.
- Python client versions are correctly bumped from 4.1.24 to 4.1.26, skipping over a missing version. Others are bumped from 4.1.25 to 4.1.26.
- In the `release-helper` run, the "release Python client" step would be launched. https://github.com/rzats/delphi-epidata/actions/runs/9616272576

dshemetov commented on .github/workflows/create-release.yml:
```diff
@@ -27,10 +27,47 @@ jobs:
+        run: |
+          cd src/client/packaging/pypi
+          bump2version --allow-dirty --new-version ${{ steps.version.outputs.next_tag }} _ignored_arg_
       - name: Create pull request into prod
```
```suggestion
      - name: Create pull request into main
```

This kept confusing me (since covidcast-indicators actually has prod)

dshemetov reviewed (approved):
So I looked over your fork and its associated PRs. The overall behavior looks to be consistent with what you describe and what you and George agreed on in #1296 (namely, Python client stays loosely coupled - it doesn't update if there are no changes to `delphi_epidata.py` or contents of `.../packaging/pypi/`, but then "catches up" to the main repo version when there is a change). No further notes from me.

melange396 commented on .github/workflows/release-helper.yml:
```diff
@@ -13,6 +13,25 @@ jobs:
+        uses: dorny/paths-filter@v3
+        with:
+          filters: |
+            python_client_changed:
+            - src/client/delphi_epidata.py
+            - src/client/packaging/pypi/**
```
for readability:
```suggestion
            python_client_changed:
              - src/client/delphi_epidata.py
              - src/client/packaging/pypi/**
```

melange396 commented on .github/workflows/create-release.yml:
```diff
@@ -27,10 +27,47 @@ jobs:
+        with:
+            base: "${{ steps.base-sha.outputs.sha }}"
+            filters: |
+              python_client_changed:
+              - src/client/delphi_epidata.py
+              - src/client/packaging/pypi/**
```
for readability:
```suggestion
              python_client_changed:
                - src/client/delphi_epidata.py
                - src/client/packaging/pypi/**
```

melange396 commented on .github/workflows/create-release.yml:
```diff
@@ -27,10 +27,47 @@ jobs:
           python -m pip install bump2version
-          echo -n "::set-output name=next_tag::"
-          bump2version --list ${{ github.event.inputs.versionName }} | grep new_version | sed -r s,"^.*=",,
+          allowed_pattern="^(major|minor|patch|[0-9]+\.[0-9]+\.[0-9]+)$"
+
+          if [[ ! ${{ github.event.inputs.versionName }} =~ $allowed_pattern ]]; then
+            echo "\nInvalid version name: ${{ github.event.inputs.versionName }}"
+            exit 1
+          fi
+
+          if [[ ${{ github.event.inputs.versionName }} =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
+            # use given version number
+            NEXT_TAG="${{ github.event.inputs.versionName }}"
+          else
+            # calculate new version number based on given tag
+            NEXT_TAG=$(bump2version --dry-run --list ${{ github.event.inputs.versionName }} | grep ^new_version | sed -r s,"^.*=",,)
+          fi
```
i dont know why this didnt occur to me when we were looking at #1473, but this is less repetitive and a little tighter:
```suggestion
          if [[ ${{ github.event.inputs.versionName }} =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
            # use given version number
            NEXT_TAG="${{ github.event.inputs.versionName }}"
          elif [[ ${{ github.event.inputs.versionName }} =~ ^(major|minor|patch)$ ]]; then
            # calculate new version number based on given tag
            NEXT_TAG=$(bump2version --dry-run --list ${{ github.event.inputs.versionName }} | grep ^new_version | sed -r s,"^.*=",,)
          else
            echo "\nInvalid version name: ${{ github.event.inputs.versionName }}"
            exit 1
          fi
```

melange396 commented on .github/workflows/create-release.yml:
```diff
@@ -27,10 +27,47 @@ jobs:
+          bump2version --new-version $NEXT_TAG _ignored_arg_
+          # save version number for later
+          echo "next_tag=$NEXT_TAG" >> $GITHUB_OUTPUT
+      - name: Get main branch SHA
+        id: base-sha
+        run: echo "sha=$(git rev-parse origin/main)" >> $GITHUB_OUTPUT
+      - name: Get Python client file changes
+        id: changed-py
+        uses: dorny/paths-filter@v3
+        with:
+            base: "${{ steps.base-sha.outputs.sha }}"
```
does this work?  its more like the usage in https://github.com/cmu-delphi/covidcast-indicators/blob/a6ea00355d79a47cc09db8090f6e5121c718de0f/.github/workflows/create-release.yml#L32-L40, and more straightforward about its intent
```suggestion
      - name: Get Python client file changes
        id: changed-py
        uses: dorny/paths-filter@v3
        with:
            base: "main"
            ref: "dev"
```

melange396 commented on .github/workflows/create-release.yml:
```diff
@@ -27,10 +27,47 @@ jobs:
+        # https://github.com/c4urself/bump2version/issues/22
+        run: |
+          cd src/client/packaging/pypi
+          bump2version --allow-dirty --new-version ${{ steps.version.outputs.next_tag }} _ignored_arg_
```
comment for clarity:
```suggestion
          # this is using the new literal version number (steps.version.outputs.next_tag) output from the bump2version step above, not the workflow input (github.event.inputs.versionName) which can be relative
          bump2version --allow-dirty --new-version ${{ steps.version.outputs.next_tag }} _ignored_arg_
```

melange396 commented on .github/workflows/release-helper.yml:
```diff
@@ -13,6 +13,25 @@ jobs:
+      - name: Get Python client file changes
+        id: changed-py
+        uses: dorny/paths-filter@v3
+        with:
+          filters: |
```
the unspecified base/ref usage seems it would be comparing `dev` to `main`, which is *probably* fine, but doing something based on whats in https://github.com/cmu-delphi/covidcast-indicators/blob/a6ea00355d79a47cc09db8090f6e5121c718de0f/.github/workflows/publish-release.yml#L31-L41 might be a little safer or at least more obvious?  idk. 
```suggestion
        with:
          # when base and ref are the same, changes are detected against the most recent commit before the push.
          # see https://github.com/dorny/paths-filter#usage for details.
          base: 'main'
          ref: 'main'
          filters: |
```

rzats commented:
@melange396 @dshemetov I've updated the PR again! Also reran the tests and updated the links here: https://github.com/cmu-delphi/delphi-epidata/pull/1465#issuecomment-2183004917

melange396 commented:
It looks like you tested this with just "patch" as the argument to "Create Release"...  Did you try with "major" (or "minor") and an explicit version to make sure those do the right thing too?

rzats commented:
@melange396 I ran some tests for these just now and everything's looking OK:

`minor`: https://github.com/rzats/delphi-epidata/pull/7 `4.1.26` → `4.2.0`
`major`: https://github.com/rzats/delphi-epidata/pull/9 `4.2.0` → `5.0.0`
custom: https://github.com/rzats/delphi-epidata/pull/11 `5.0.0` → `6.7.8`; also changes the Python client to make sure it jumps to the right version
